server/src/main/java/io/crate/planner/GCDanglingArtifactsPlan.java

```diff
@@ -0,0 +1,90 @@
+ * additional information regarding copyright ownership.  Crate licenses
+ * this file to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.planner;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import org.elasticsearch.action.support.master.AcknowledgedResponse;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.cluster.metadata.Metadata;
+import org.elasticsearch.cluster.metadata.RelationMetadata;
+import org.elasticsearch.index.Index;
+
+import com.carrotsearch.hppc.cursors.ObjectCursor;
+
+import io.crate.data.Row;
+import io.crate.data.RowConsumer;
+import io.crate.execution.support.OneRowActionListener;
+import io.crate.planner.operators.SubQueryResults;
+
+public final class GCDanglingArtifactsPlan implements Plan {
+
+    @Override
+    public StatementType type() {
+        return StatementType.MANAGEMENT;
+    }
+
+    @Override
+    public void executeOrFail(DependencyCarrier dependencies,
+                              PlannerContext plannerContext,
+                              RowConsumer consumer,
+                              Row params,
+                              SubQueryResults subQueryResults) {
+        var listener = OneRowActionListener.oneIfAcknowledged(consumer);
+        try {
+            ClusterState clusterState = dependencies.clusterService().state();
+            Metadata metadata = clusterState.metadata();
+
+            Set<Index> danglingIndicesToDelete = new HashSet<>();
+            Set<Index> allTableIndices = new HashSet<>();
+            for (RelationMetadata.Table table : metadata.tableRelations()) {
+                allTableIndices.addAll(metadata.getIndices(
+                    table.name(),
+                    List.of(),
+                    false,
+                    IndexMetadata::getIndex)
+                );
+            }
+            for (RelationMetadata.BlobTable blob : metadata.blobRelations()) {
+                allTableIndices.addAll(metadata.getIndices(
+                    blob.name(),
+                    List.of(),
+                    false,
+                    IndexMetadata::getIndex)
+                );
+            }
```

Comment:
I'd tend to extend `getIndices` to return an empty list for cases where there is a relation which is neither Table/BlobTable - so the code here doesn't need to distinguish table/blobTable but can just do `metadata.relations(RelationMetadata.class)` together with the other suggested change.

Reduce the amount of areas where we special case the concrete `RelationMetadata` instances.